File: python3/Solution438.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        res = []
        s_len = len(s)
        p_len = len(p)
        if(p_len > s_len):
            return res
        count = [0]*26
        zeros = [0]*26
        for i in range(p_len):
            count[ord(p[i])-97]+=1
            count[ord(s[i])-97]-=1
        if count == zeros:
            res.append(0)
        ps = 0
        pf = p_len
        while pf < s_len:
            count[ord(s[pf])-97]-=1
            count[ord(s[ps])-97]+=1
            pf+=1
            ps+=1
            if count == zeros:
                res.append(ps)
        return res

File: java/Solution438.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> res = new ArrayList<>();
        int ls = s.length();
        int lp = p.length();
        if(lp > ls) return res;
        int[] count = new int[26];
        char[] cs = s.toCharArray();
        char[] cp = p.toCharArray();
        for(char c: cp){
            count[c-'a']++;
        }
        int ps = 0, pf = 0;
        for(pf = 0; pf< lp; pf++){
            count[ cs[pf]-'a' ]--;
        }
        if(check(count)) res.add(ps);
        while(pf< ls){
            count[ cs[ps++]-'a'] ++;
            count[ cs[pf++]-'a'] --;
            if(check(count)) res.add(ps);
        }
        return res;
    }
    private boolean check(int[] a){
        for (int k=0; k<26; k++){
                if(a[k] != 0) return false;
        }
        return true;
    }
}
